src/operation.rs

```diff
@@ -473,6 +492,68 @@ impl Operation {
                     .send(Message::PendingProgress(id, 100.0))
                     .await;
             }
+            Self::Extract { paths, to } => {
+                for path in paths {
+                    let to = to.to_owned();
+
+                    tokio::task::spawn_blocking(move || -> Result<(), Error> {
+                        let mut to = path.file_stem().map(|stem| to.join(stem)).unwrap();
+
+                        if to.exists() {
+                            let mut extensionless_path = path.to_owned();
+                            extensionless_path.set_extension("");
+                            to = copy_unique_path(&extensionless_path, &to.parent().unwrap());
+                        };
+
+                        fs::create_dir_all(&to)?;
+
+                        let mut archive = fs::File::open(path)
```

Comment:
It would make sense if the extractor is in the separate function, the level of nesting makes this hard to review even on a large screen.